File: skills/kalshi-trading/scripts/kalshi-client.ts
#!/usr/bin/env npx tsx
/**
 * Kalshi API Client
 *
 * A ready-to-use client for interacting with Kalshi prediction markets.
 * Can be executed directly with `npx tsx` or imported as a module.
 *
 * Usage:
 *   # As a script - list open markets
 *   npx tsx kalshi-client.ts markets --limit 10
 *
 *   # As a script - get specific market
 *   npx tsx kalshi-client.ts market PRES-2024-DEM
 *
 *   # As a module
 *   import { KalshiClient } from './kalshi-client';
 *   const client = new KalshiClient();
 *   const markets = await client.getMarkets({ limit: 10 });
 */

import * as crypto from "crypto";

// ==================== Types ====================

export interface KalshiClientConfig {
  apiKey?: string;
  privateKeyPem?: string;
  baseUrl?: string;
  demo?: boolean;
}

export interface Market {
  ticker: string;
  title: string;
  status: string;
  yes_bid?: number;
  yes_ask?: number;
  no_bid?: number;
  no_ask?: number;
  volume?: number;
  open_interest?: number;
  close_time?: string;
  [key: string]: unknown;
}

export interface Orderbook {
  yes: [number, number][];
  no: [number, number][];
}

export interface Trade {
  trade_id: string;
  ticker: string;
  count: number;
  yes_price: number;
  no_price: number;
  taker_side: string;
  created_time: string;
}

export interface Event {
  event_ticker: string;
  title: string;
  category?: string;
  markets?: string[];
  [key: string]: unknown;
}

export interface Balance {
  balance: number;
  payout: number;
}

export interface Position {
  ticker: string;
  position: number;
  market_exposure: number;
  realized_pnl: number;
  total_traded: number;
  resting_orders_count: number;
}

export interface Order {
  order_id: string;
  ticker: string;
  side: "yes" | "no";
  action: "buy" | "sell";
  type: "limit" | "market";
  yes_price?: number;
  no_price?: number;
  count: number;
  remaining_count: number;
  status: string;
  created_time: string;
}

export interface CreateOrderParams {
  ticker: string;
  side: "yes" | "no";
  action: "buy" | "sell";
  count: number;
  price: number;
  type?: "limit" | "market";
}

// ==================== Client ====================

export class KalshiClient {
  private readonly apiKey?: string;
  private readonly privateKeyPem?: string;
  private readonly baseUrl: string;

  static readonly DEFAULT_BASE_URL =
    "https://api.elections.kalshi.com/trade-api/v2";
  static readonly DEMO_BASE_URL = "https://demo-api.kalshi.co/trade-api/v2";

  constructor(config: KalshiClientConfig = {}) {
    this.apiKey = config.apiKey ?? process.env.KALSHI_API_KEY;
    this.privateKeyPem = config.privateKeyPem ?? process.env.KALSHI_PRIVATE_KEY;

    if (config.demo) {
      this.baseUrl = KalshiClient.DEMO_BASE_URL;
    } else {
      this.baseUrl = config.baseUrl ?? KalshiClient.DEFAULT_BASE_URL;
    }
  }

  private getAuthHeaders(method: string, path: string): Record<string, string> {
    if (!this.apiKey || !this.privateKeyPem) {
      throw new Error(
        "API key and private key required for authenticated requests"
      );
    }

    const timestamp = Date.now().toString();
    const message = `${timestamp}${method}${path}`;

    const sign = crypto.createSign("RSA-SHA256");
    sign.update(message);
    const signature = sign.sign(this.privateKeyPem, "base64");

    return {
      "KALSHI-ACCESS-KEY": this.apiKey,
      "KALSHI-ACCESS-SIGNATURE": signature,
      "KALSHI-ACCESS-TIMESTAMP": timestamp,
      "Content-Type": "application/json",
    };
  }

  private async request<T>(
    method: string,
    path: string,
    options: {
      params?: Record<string, string | number>;
      body?: Record<string, unknown>;
      authenticated?: boolean;
    } = {}
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}${path}`);

    if (options.params) {
      Object.entries(options.params).forEach(([key, value]) => {
        url.searchParams.set(key, String(value));
      });
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };

    if (options.authenticated) {
      Object.assign(headers, this.getAuthHeaders(method, path));
    }

    const response = await fetch(url.toString(), {
      method,
      headers,
      body: options.body ? JSON.stringify(options.body) : undefined,
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Kalshi API error: ${response.status} - ${error}`);
    }

    return response.json();
  }

  // ==================== Public Endpoints ====================

  async getMarkets(options: {
    limit?: number;
    status?: string;
    cursor?: string;
    eventTicker?: string;
  } = {}): Promise<Market[]> {
    const params: Record<string, string | number> = {
      limit: options.limit ?? 100,
      status: options.status ?? "open",
    };
    if (options.cursor) params.cursor = options.cursor;
    if (options.eventTicker) params.event_ticker = options.eventTicker;

    const result = await this.request<{ markets: Market[] }>("GET", "/markets", {
      params,
    });
    return result.markets ?? [];
  }

  async getMarket(ticker: string): Promise<Market> {
    const result = await this.request<{ market: Market }>(
      "GET",
      `/markets/${ticker}`
    );
    return result.market;
  }

  async getOrderbook(ticker: string, depth?: number): Promise<Orderbook> {
    const params: Record<string, string | number> = {};
    if (depth) params.depth = depth;

    const result = await this.request<{ orderbook: Orderbook }>(
      "GET",
      `/markets/${ticker}/orderbook`,
      { params }
    );
    return result.orderbook;
  }

  async getTrades(
    ticker: string,
    options: { limit?: number; cursor?: string } = {}
  ): Promise<Trade[]> {
    const params: Record<string, string | number> = {
      limit: options.limit ?? 100,
    };
    if (options.cursor) params.cursor = options.cursor;

    const result = await this.request<{ trades: Trade[] }>(
      "GET",
      `/markets/${ticker}/trades`,
      { params }
    );
    return result.trades ?? [];
  }

  async getEvents(options: {
    limit?: number;
    status?: string;
    cursor?: string;
  } = {}): Promise<Event[]> {
    const params: Record<string, string | number> = {
      limit: options.limit ?? 100,
      status: options.status ?? "open",
    };
    if (options.cursor) params.cursor = options.cursor;

    const result = await this.request<{ events: Event[] }>("GET", "/events", {
      params,
    });
    return result.events ?? [];
  }

  async getEvent(eventTicker: string): Promise<Event> {
    const result = await this.request<{ event: Event }>(
      "GET",
      `/events/${eventTicker}`
    );
    return result.event;
  }

  // ==================== Authenticated Endpoints ====================

  async getBalance(): Promise<Balance> {
    return this.request<Balance>("GET", "/portfolio/balance", {
      authenticated: true,
    });
  }

  async getPositions(options: {
    ticker?: string;
    settlementStatus?: string;
  } = {}): Promise<Position[]> {
    const params: Record<string, string | number> = {
      settlement_status: options.settlementStatus ?? "unsettled",
    };
    if (options.ticker) params.ticker = options.ticker;

    const result = await this.request<{ market_positions: Position[] }>(
      "GET",
      "/portfolio/positions",
      { params, authenticated: true }
    );
    return result.market_positions ?? [];
  }

  async getOrders(options: {
    ticker?: string;
    status?: string;
  } = {}): Promise<Order[]> {
    const params: Record<string, string | number> = {
      status: options.status ?? "resting",
    };
    if (options.ticker) params.ticker = options.ticker;

    const result = await this.request<{ orders: Order[] }>(
      "GET",
      "/portfolio/orders",
      { params, authenticated: true }
    );
    return result.orders ?? [];
  }

  async createOrder(params: CreateOrderParams): Promise<Order> {
    const body: Record<string, unknown> = {
      ticker: params.ticker,
      side: params.side,
      action: params.action,
      type: params.type ?? "limit",
      count: params.count,
    };

    if (body.type === "limit") {
      if (params.side === "yes") {
        body.yes_price = params.price;
      } else {
        body.no_price = params.price;
      }
    }

    const result = await this.request<{ order: Order }>(
      "POST",
      "/portfolio/orders",
      { body, authenticated: true }
    );
    return result.order;
  }

  async cancelOrder(orderId: string): Promise<Order> {
    const result = await this.request<{ order: Order }>(
      "DELETE",
      `/portfolio/orders/${orderId}`,
      { authenticated: true }
    );
    return result.order;
  }
}

// ==================== CLI ====================

async function main() {
  const args = process.argv.slice(2);
  const command = args[0];

  if (!command) {
    console.log(`
Kalshi API Client

Usage:
  npx tsx kalshi-client.ts <command> [options]

Commands:
  markets [--limit N] [--status open|closed]  List markets
  market <ticker>                             Get market details
  orderbook <ticker>                          Get orderbook
  balance                                     Get account balance (requires auth)
  positions                                   Get positions (requires auth)
    `);
    return;
  }

  const client = new KalshiClient();

  switch (command) {
    case "markets": {
      const limitIdx = args.indexOf("--limit");
      const limit = limitIdx !== -1 ? parseInt(args[limitIdx + 1]) : 10;
      const statusIdx = args.indexOf("--status");
      const status = statusIdx !== -1 ? args[statusIdx + 1] : "open";

      const markets = await client.getMarkets({ limit, status });
      for (const market of markets) {
        console.log(`${market.ticker}: ${market.title}`);
        console.log(
          `  Yes: ${market.yes_bid ?? "N/A"}¢ - ${market.yes_ask ?? "N/A"}¢`
        );
        console.log();
      }
      break;
    }

    case "market": {
      const ticker = args[1];
      if (!ticker) {
        console.error("Usage: market <ticker>");
        process.exit(1);
      }
      const market = await client.getMarket(ticker);
      console.log(JSON.stringify(market, null, 2));
      break;
    }

    case "orderbook": {
      const ticker = args[1];
      if (!ticker) {
        console.error("Usage: orderbook <ticker>");
        process.exit(1);
      }
      const orderbook = await client.getOrderbook(ticker);
      console.log(JSON.stringify(orderbook, null, 2));
      break;
    }

    case "balance": {
      const balance = await client.getBalance();
      console.log(`Balance: $${(balance.balance / 100).toFixed(2)}`);
      break;
    }

    case "positions": {
      const positions = await client.getPositions();
      for (const pos of positions) {
        console.log(`${pos.ticker}: ${pos.position} contracts`);
      }
      break;
    }

    default:
      console.error(`Unknown command: ${command}`);
      process.exit(1);
  }
}

// Run CLI if executed directly
main().catch((error) => {
  console.error("Error:", error.message);
  process.exit(1);
});

File: skills/kalshi-trading/scripts/kalshi_client.py
#!/usr/bin/env python3
"""
Kalshi API Client

A ready-to-use client for interacting with Kalshi prediction markets.
This script can be executed directly or imported as a module.

Usage:
    # As a script - list open markets
    python kalshi_client.py markets --limit 10
    
    # As a script - get specific market
    python kalshi_client.py market PRES-2024-DEM
    
    # As a module
    from kalshi_client import KalshiClient
    client = KalshiClient()
    markets = client.get_markets(limit=10)
"""

import os
import sys
import time
import json
import base64
import argparse
from typing import Optional, Dict, List, Any

try:
    import requests
    from cryptography.hazmat.primitives import hashes, serialization
    from cryptography.hazmat.primitives.asymmetric import padding
except ImportError:
    print("Required packages not installed. Run:")
    print("  pip install requests cryptography")
    sys.exit(1)


class KalshiClient:
    """Client for Kalshi Trade API v2."""
    
    DEFAULT_BASE_URL = "https://api.elections.kalshi.com/trade-api/v2"
    DEMO_BASE_URL = "https://demo-api.kalshi.co/trade-api/v2"
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        private_key_pem: Optional[str] = None,
        base_url: Optional[str] = None,
        demo: bool = False
    ):
        """
        Initialize Kalshi client.
        
        Args:
            api_key: Kalshi API key ID (or set KALSHI_API_KEY env var)
            private_key_pem: RSA private key in PEM format (or set KALSHI_PRIVATE_KEY)
            base_url: API base URL (default: production)
            demo: Use demo environment instead of production
        """
        self.api_key = api_key or os.environ.get("KALSHI_API_KEY")
        self.private_key_pem = private_key_pem or os.environ.get("KALSHI_PRIVATE_KEY")
        
        if demo:
            self.base_url = self.DEMO_BASE_URL
        else:
            self.base_url = base_url or self.DEFAULT_BASE_URL
        
        self._private_key = None
        if self.private_key_pem:
            self._private_key = serialization.load_pem_private_key(
                self.private_key_pem.encode(),
                password=None
            )
    
    def _get_auth_headers(self, method: str, path: str) -> Dict[str, str]:
        """Generate authenticated headers."""
        if not self.api_key or not self._private_key:
            raise ValueError("API key and private key required for authenticated requests")
        
        timestamp = str(int(time.time() * 1000))
        message = f"{timestamp}{method}{path}"
        
        signature = self._private_key.sign(
            message.encode(),
            padding.PKCS1v15(),
            hashes.SHA256()
        )
        
        return {
            "KALSHI-ACCESS-KEY": self.api_key,
            "KALSHI-ACCESS-SIGNATURE": base64.b64encode(signature).decode(),
            "KALSHI-ACCESS-TIMESTAMP": timestamp,
            "Content-Type": "application/json"
        }
    
    def _request(
        self,
        method: str,
        path: str,
        params: Optional[Dict] = None,
        json_data: Optional[Dict] = None,
        authenticated: bool = False
    ) -> Dict[str, Any]:
        """Make API request."""
        url = f"{self.base_url}{path}"
        headers = {"Content-Type": "application/json"}
        
        if authenticated:
            headers = self._get_auth_headers(method, path)
        
        response = requests.request(
            method=method,
            url=url,
            headers=headers,
            params=params,
            json=json_data
        )
        
        response.raise_for_status()
        return response.json()
    
    # ==================== Public Endpoints ====================
    
    def get_markets(
        self,
        limit: int = 100,
        status: str = "open",
        cursor: Optional[str] = None,
        event_ticker: Optional[str] = None
    ) -> List[Dict]:
        """Get list of markets."""
        params = {"limit": limit, "status": status}
        if cursor:
            params["cursor"] = cursor
        if event_ticker:
            params["event_ticker"] = event_ticker
        
        result = self._request("GET", "/markets", params=params)
        return result.get("markets", [])
    
    def get_market(self, ticker: str) -> Dict:
        """Get details for a specific market."""
        result = self._request("GET", f"/markets/{ticker}")
        return result.get("market", {})
    
    def get_orderbook(self, ticker: str, depth: Optional[int] = None) -> Dict:
        """Get orderbook for a market."""
        params = {}
        if depth:
            params["depth"] = depth
        result = self._request("GET", f"/markets/{ticker}/orderbook", params=params)
        return result.get("orderbook", {})
    
    def get_trades(
        self,
        ticker: str,
        limit: int = 100,
        cursor: Optional[str] = None
    ) -> List[Dict]:
        """Get recent trades for a market."""
        params = {"limit": limit}
        if cursor:
            params["cursor"] = cursor
        result = self._request("GET", f"/markets/{ticker}/trades", params=params)
        return result.get("trades", [])
    
    def get_events(
        self,
        limit: int = 100,
        status: str = "open",
        cursor: Optional[str] = None
    ) -> List[Dict]:
        """Get list of events."""
        params = {"limit": limit, "status": status}
        if cursor:
            params["cursor"] = cursor
        result = self._request("GET", "/events", params=params)
        return result.get("events", [])
    
    def get_event(self, event_ticker: str) -> Dict:
        """Get details for a specific event."""
        result = self._request("GET", f"/events/{event_ticker}")
        return result.get("event", {})
    
    # ==================== Authenticated Endpoints ====================
    
    def get_balance(self) -> Dict:
        """Get account balance (requires authentication)."""
        return self._request("GET", "/portfolio/balance", authenticated=True)
    
    def get_positions(
        self,
        ticker: Optional[str] = None,
        settlement_status: str = "unsettled"
    ) -> List[Dict]:
        """Get current positions (requires authentication)."""
        params = {"settlement_status": settlement_status}
        if ticker:
            params["ticker"] = ticker
        result = self._request("GET", "/portfolio/positions", params=params, authenticated=True)
        return result.get("market_positions", [])
    
    def get_orders(
        self,
        ticker: Optional[str] = None,
        status: str = "resting"
    ) -> List[Dict]:
        """Get orders (requires authentication)."""
        params = {"status": status}
        if ticker:
            params["ticker"] = ticker
        result = self._request("GET", "/portfolio/orders", params=params, authenticated=True)
        return result.get("orders", [])
    
    def create_order(
        self,
        ticker: str,
        side: str,
        action: str,
        count: int,
        price: int,
        order_type: str = "limit"
    ) -> Dict:
        """
        Create an order (requires authentication).
        
        Args:
            ticker: Market ticker
            side: "yes" or "no"
            action: "buy" or "sell"
            count: Number of contracts
            price: Price in cents (1-99)
            order_type: "limit" or "market"
        """
        payload = {
            "ticker": ticker,
            "side": side,
            "action": action,
            "type": order_type,
            "count": count
        }
        
        if order_type == "limit":
            if side == "yes":
                payload["yes_price"] = price
            else:
                payload["no_price"] = price
        
        result = self._request("POST", "/portfolio/orders", json_data=payload, authenticated=True)
        return result.get("order", {})
    
    def cancel_order(self, order_id: str) -> Dict:
        """Cancel an order (requires authentication)."""
        result = self._request("DELETE", f"/portfolio/orders/{order_id}", authenticated=True)
        return result.get("order", {})


def main():
    """CLI entry point."""
    parser = argparse.ArgumentParser(description="Kalshi API Client")
    subparsers = parser.add_subparsers(dest="command", help="Available commands")
    
    # Markets command
    markets_parser = subparsers.add_parser("markets", help="List markets")
    markets_parser.add_argument("--limit", type=int, default=10, help="Number of results")
    markets_parser.add_argument("--status", default="open", help="Market status filter")
    
    # Market command
    market_parser = subparsers.add_parser("market", help="Get market details")
    market_parser.add_argument("ticker", help="Market ticker")
    
    # Orderbook command
    orderbook_parser = subparsers.add_parser("orderbook", help="Get orderbook")
    orderbook_parser.add_argument("ticker", help="Market ticker")
    
    # Balance command
    subparsers.add_parser("balance", help="Get account balance (requires auth)")
    
    # Positions command
    subparsers.add_parser("positions", help="Get positions (requires auth)")
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return
    
    client = KalshiClient()
    
    if args.command == "markets":
        result = client.get_markets(limit=args.limit, status=args.status)
        for market in result:
            print(f"{market['ticker']}: {market['title']}")
            print(f"  Yes: {market.get('yes_bid', 'N/A')}¢ - {market.get('yes_ask', 'N/A')}¢")
            print()
    
    elif args.command == "market":
        result = client.get_market(args.ticker)
        print(json.dumps(result, indent=2))
    
    elif args.command == "orderbook":
        result = client.get_orderbook(args.ticker)
        print(json.dumps(result, indent=2))
    
    elif args.command == "balance":
        result = client.get_balance()
        balance_dollars = result.get("balance", 0) / 100
        print(f"Balance: ${balance_dollars:.2f}")
    
    elif args.command == "positions":
        result = client.get_positions()
        for pos in result:
            print(f"{pos['ticker']}: {pos['position']} contracts")


if __name__ == "__main__":
    main()

